drivers: udc: initial support for USB Device Controller on Renesas RA family

thenguyenyf opened the pull request:
Support USB device_next stack on Renesas RA family
Add support for USBHS port on these boards:
- ek_ra8m1
- ek_ra8d1
- ek_ra6m3
- ek_ra6m5

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+
+#define DT_DRV_COMPAT renesas_ra_udc
+
+#define USBHS_BASE_ADDR         (0x40252000)
+#define USB_MODULE_NUMBER(addr) ((addr == USBHS_BASE_ADDR) ? 1 : 0)
```
I did not understand the reason for this construction.

thenguyenyf replied:
This code is to determine the module number of USB before input to HAL layer. The Renesas RA device has two USB modules, USBFS and USBHS, with module numbers 0 and 1 respectively.

jfischer-no replied:
If there are different controllers, should not there be two different drivers?

thenguyenyf replied:
There is not much hardware difference between USBHS and USBFS, so they can share the same driver source. Actually, on other platforms (like Renesas RA FSP or TinyUSB library), only 1 driver source is needed for 2 modules.

jfischer-no replied:
I found out how module_number is used in your HAL.

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+	static struct udc_renesas_ra_data udc_priv_##n = {                                         \
+		.udc_cfg =                                                                         \
+			{                                                                          \
+				.module_number =                                                   \
```
```c

		.udc_cfg = {                                                                          \
			.module_number =   
```

thenguyenyf replied:
It seems you mentioned a code formatting issue. But I ran the clang-format tool here and it made the code look like this.

jfischer-no replied:
Well, your code does not look right and you need to fix it.

thenguyenyf replied:
I fixed it

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+void usb_interrupt_handler(void *arg)
+{
+	ARG_UNUSED(arg);
+	usb_device_isr();
```
The `usb_device_isr` is implemented in the Renesas HAL module as a low-level driven routine with direct access to registers. Other processing for UDC functionality is still handled in the shim driver, based on hardware events that this ISR provides via callbacks. It seems our implementation is quite similar to others that have external modules for the HAL layer, such as NXP in `drivers/serial/uart_nxp_s32_linflexd.c` or STM in `drivers/usb/udc/udc_stm32.c`.

jfischer-no replied:
> It seems our implementation is quite similar to others that have external modules for the HAL layer, such as NXP in `drivers/serial/uart_nxp_s32_linflexd.c` or STM in `drivers/usb/udc/udc_stm32.c`.

No your implementation is not similar to `drivers/usb/udc/udc_stm32.c`. I do not understand why you are referring to the NXP UART driver. Your shim driver design does not look good, and you will probably have problems with controllers that support high speed.

thenguyenyf replied:
I mean I see this as a common way when a shim driver, not only USB driver but also other drivers, uses the HAL API to implement its functionality. In that case, the HAL ISR just handle the low-level hardware routines when an interrupt occurs, get the interrupt event and notify the upper layer by callback.

> and you will probably have problems with controllers that support high speed.

For this comment, could you give me more details about your concern? Did you mention that it might cause this driver to have low performance and not be able to adapt to high-speed USB? Actually, I have tested this driver can work well as a high-speed device with currently supported samples (cdc_acm, mass, hid, ...)

jfischer-no replied:
For example https://github.com/zephyrproject-rtos/zephyr/pull/79989#discussion_r1805420830

thenguyenyf replied:
> For example [#79989 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79989#discussion_r1805420830)

I agreed with this comment and already have an update for it in local. I mean what about this topic? I think we are not the first to implement low level ISR somewhere other than shim driver, but in HAL module.

jfischer-no replied:
Please push your changes.

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+#if !USBHS_PHY_CLOCK_SOURCE_IS_XTAL
+	if (data->udc_cfg.usb_speed == USBD_SPEED_HS) {
+		LOG_ERR("High-speed operation is not supported in case PHY clock source is not "
+			"XTAL");
+		return -ENOTSUP;
+	}
+
+	uint32_t uclk_src = RA_CGC_CLK_SRC(DT_CLOCKS_CTLR(DT_NODELABEL(uclk)));
+	uint32_t uclk_div = DT_PROP_OR(DT_NODELABEL(uclk), div, 1);
+	uint32_t u60clk_src = RA_CGC_CLK_SRC(DT_CLOCKS_CTLR(DT_NODELABEL(u60clk)));
+	uint32_t u60clk_div = DT_PROP_OR(DT_NODELABEL(u60clk), div, 1);
+
+	if (uclk_src == BSP_CLOCKS_CLOCK_DISABLED || u60clk_src == BSP_CLOCKS_CLOCK_DISABLED) {
+		LOG_ERR("PHY clock is not working");
+		return -EINVAL;
+	}
+
+	uint32_t uclk_clock_rate = R_BSP_SourceClockHzGet(uclk_src) / uclk_div;
+	uint32_t u60clk_clock_rate = R_BSP_SourceClockHzGet(u60clk_src) / u60clk_div;
+
+	if (uclk_clock_rate != 48000000) {
+		LOG_ERR("Setting for uclk should be 48Mhz");
+		return -ENOTSUP;
+	}
+
+	if (u60clk_clock_rate != 60000000) {
+		LOG_ERR("Setting for u60clk should be 60Mhz");
+		return -ENOTSUP;
+	}
```
Renesas RA clock_control does not yet support the input clock of a physical node, only that of a device driver node. So, I have this workaround before the feature for a physical node clock is added.

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+				udc_submit_event(dev, UDC_EVT_RESUME, 0);
+				break;
+
+			case USBD_EVENT_SOF:
```
I fixed it

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,754 @@
+	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
+}
+
+void usb_interrupt_handler(void *arg)
```
I fixed it

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+{
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	k_fifo_put(&cfg->fifo, buf);
+
+	return FSP_SUCCESS == R_USBD_XferStart(&data->udc, cfg->addr, buf->data, buf->size) ? 0
+											    : -EIO;
+}
```
```c
	fsp_err_t err;

...

	err = R_USBD_XferStart(&data->udc, cfg->addr, buf->data, buf->size);
	if (err != FSP_SUCCESS) {
		return -EIO;
	}

	return 0;
```

thenguyenyf replied:
I fixed it

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+
+static void udc_event_xfer_complete(const struct device *dev, struct udc_renesas_ra_evt *evt)
+{
+	struct net_buf *buf;
+
+	unsigned int ep = evt->hal_evt.xfer_complete.ep_addr;
+	unsigned int result = evt->hal_evt.xfer_complete.result;
+	unsigned int len = evt->hal_evt.xfer_complete.len;
```
```c
uint8_t ep = evt->hal_evt.xfer_complete.ep_addr;
usbd_xfer_result_t result = evt->hal_evt.xfer_complete.result;
uint32_t len = evt->hal_evt.xfer_complete.len;
```
?

thenguyenyf replied:
I fixed it

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+
+		case UDC_RENESAS_RA_EVT_STATUS:
+			udc_event_status_in(dev);
+
```
Maybe `__fallthrough;`

thenguyenyf replied:
I fixed it

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+		return 0;
+	}
+
+	struct udc_renesas_ra_evt evt = {
+		.type = UDC_RENESAS_RA_EVT_XFER,
+		.ep = cfg->addr,
+	};
```
Variable definitions should be at the top of a block. Please fix it everywhere.

thenguyenyf replied:
I fixed it

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+	struct net_buf *buf;
+
+	lock_key = irq_lock();
+
```
I guess you need to cancel the transfer in the HAL driver.

jfischer-no commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+	if (buf != NULL) {
+		int err;
+
+		if (USB_EP_DIR_IS_IN(ep)) {
```
You need to check udc_ep_buf_has_zlp(buf), see how it is used in existing drivers.

thenguyenyf replied:
I fixed it

jfischer-no commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -527,6 +527,25 @@
 				status = "disabled";
 			};
 		};
+
```
Please put dts and board related changes into separate commits. Your PR would have four commits, manifest change, dts changes, driver implementation, board-related changes.

thenguyenyf replied:
I fixed it

jfischer-no commented:
@kasjer Please review.

kasjer commented:
> @kasjer Please review.

review in progress...

kasjer commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+	}
+
+	if (result == USBD_XFER_RESULT_SUCCESS) {
+		udc_ep_set_busy(dev, ep, 0);
```
Comment from @jfischer-no about ZLP could be potentially solved here
```c
		/* ZLP needed? */
		if (USB_EP_DIR_IS_IN(ep) && udc_ep_buf_has_zlp(buf)) {
			udc_ep_buf_clear_zlp(buf);
			/* Send ZLP, notification about transfer complete should come again */
			... = R_USBD_XferStart(udc, USB_CONTROL_EP_IN, NULL, 0);
			return;
		}
```

To test that it works, code could make some string descriptor that will have 64 bytes total. Maybe manufacturer name or serial number. This case then could be tested during enumeration.

**But udc_buf_get() should be change to udc_buf_peek() for ZLP !!!**

thenguyenyf replied:
I fixed it

tmon-nordic replied:
Control IN transfers do have ZLP during Data Stage only if the payload is multiple of `bMaxPacketSize0` and the host requested `wLength` greater than multiple of `bMaxPacketSize0`.

kasjer commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	/* Clear setup endpoint. Send a ZLP */
+	R_USBD_XferStart(&data->udc, USB_CONTROL_EP_IN, NULL, 0);
```
Are you sure that sending ZLP for every control transfer is needed?

thenguyenyf replied:
I fixed it

kasjer commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup_packet, sizeof(struct usb_setup_packet));
+	net_buf_add(buf, sizeof(struct usb_setup_packet));
```
maybe
```c
net_buf_add_mem(buf, setup_packet, sizeof(struct usb_setup_packet));
```

thenguyenyf replied:
I don't really understand that case. You suggest that if `net_buf_add_mem` is used, we could avoid using the memcpy and save some CPU cycles, is this right? Or just a way to make code prettier?

kasjer replied:
Just prettier, it combines to line into one

kasjer reviewed (commented):
Not tested, just code review

thenguyenyf commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -527,6 +527,25 @@
+			reg = <0x40252000 0x2000>;
+			interrupts = <54 12>, <55 12>, <56 12>;
+			interrupt-names = "usbhs-ir", "usbhs-d0", "usbhs-d1";
+			num-bidir-endpoints = <10>;
```
Can this prop be used for user input to select the device's operating mode? Or is it a device capability that indicates the highest speed the device controller can run at?
If user can config it, I will move this config from RA6x SoC to board layer, same way as RA8

kasjer replied:
I think that it can be skipped. I noticed a difference, but now I think it does not matter so disregard this comment.

thenguyenyf commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -527,6 +527,25 @@
+		usbhs: usbhs@40252000 {
+			compatible = "renesas,ra-usb";
+			reg = <0x40252000 0x2000>;
+			interrupts = <54 12>, <55 12>, <56 12>;
```
Do you expect higher priority for RA8? I have tested this interrupt configuration using the available sample and it can run well.

kasjer replied:
I noticed difference, I don't know if it matters, if it works I don't object to this difference. I just wanted to let you know that there may be something there.

thenguyenyf committed:
boards: renesas: add support for USB device on Renesas RA boards

Add support for UDC on highspeed port on these boards:
- ek_ra8m1
- ek_ra8d1
- ek_ra6m5
- ek_ra6m3

Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+
+extern void usb_device_isr(void);
+
+static void udc_event_handler(usbd_callback_arg_t *p_args)
```
I fixed it

thenguyenyf commented on drivers/usb/udc/udc_renesas_ra.c:
```diff
@@ -0,0 +1,756 @@
+	}
+}
+
+static void usb_interrupt_handler(void *arg)
```
I fixed it

kasjer reviewed (approved):
LGTM

thenguyenyf commented:
I rebase main branch to solve conflict

thenguyenyf commented:
Last push to rebase main branch and solve conflict

thenguyenyf commented:
Last push to update the manifest rev because the manifest update was merged

thenguyenyf commented:
Hi @KhiemNguyenT , @soburi , @thaoluonguw , @duynguyenxa . Could you please help to take a look?

thenguyenyf committed:
drivers: udc: add support for Renesas RA USB device

First commit to support UDC on Renesas RA USBHS module

Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>

thenguyenyf committed:
dts: arm: renesas: add support for UDC on Renesas RA SoC

Add device node to support UDC driver on these SoC: RA6M3,RA6M5,
RA8M1, RA8D1

Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>

thenguyenyf commented:
Last push to update for dts information and correct typos

thaoluonguw reviewed (dismissed):
It looks good to me. Thank you.

thenguyenyf commented:
Hi @KhiemNguyenT , @soburi , @duynguyenxa . Could you please check it?

fabiobaltieri commented:
@thaoluonguw hey this needs a rebase

thenguyenyf commented:
> @thaoluonguw hey this needs a rebase

Hi @fabiobaltieri . I have rebased main branch and solve the conflict.

fabiobaltieri commented:
@jfischer-no can you take a look?

fabiobaltieri commented:
@jfischer-no can you take a look?

thenguyenyf commented:
Last push to rebase main and solve conflict

thaoluonguw commented on dts/arm/renesas/ra/ra8/r7fa8m1xh.dtsi:
```diff
@@ -241,4 +241,25 @@
+	soc {
+		usbhs: usbhs@40252000 {
+			compatible = "renesas,ra-usb";
+			reg = <0x40252000 0x2000>;
```
It should be "0x40351000"

thenguyenyf replied:
I fixed it

jfischer-no commented on boards/renesas/ek_ra6m3/doc/index.rst:
```diff
@@ -102,6 +102,8 @@ The below features are currently supported on Zephyr OS for EK-RA6M3 board:
 | COUNTER   | on-chip    | counter              |
 +-----------+------------+----------------------+
+| USBHS     | on-chip    | udc                  |
++-----------+------------+----------------------+
```
@thenguyenyf Please add test feature `usbd` to board's yaml files, example https://github.com/zephyrproject-rtos/zephyr/blob/e1b43207224e2226dc692bd294afaf869e05a201/boards/renesas/da14695_dk_usb/da14695_dk_usb.yaml#L17 In this or a subsequent PR.
